# DATA ANALYSIS EXERCISE

## --------------------------------------------------------------------------------------------------------------------------------
## Imports and reading in the jsonlines file

In [299]:
# Required imports 
import json
import re
import pandas as pd
from pandas.io.json import json_normalize

In [300]:
# Read in the jsonl file
data_file = 'ida_wrangling_exercise_data.2017-02-13.jsonl'

with open(data_file) as f:
    flines = f.readlines()

## json_normalize to flatten the JSON objects into a pandas dataframe

In [301]:
data = []
for line in flines:
    data.append(json.loads(line))

df = json_normalize(data)

## --------------------------------------------------------------------------------------------------------------------------------
## Basic Exploratory Analysis

1. Size of the dataset: df.shape
2. Top n rows: df.head(n)
3. Last n rows: df.tail(n)
4. n Random sample: df.sample(n)
5. Description: df.describe()

In [302]:
# Total rows and columns 
rows, columns = df.shape
df.shape

(150000, 15)

In [303]:
# Top 5 rows
df.head(5)

address     address.city  \
0                                               NaN        Hoodburgh   
1  20722 Coleman Villages\nEast Rose, SC 71064-5894              NaN   
2                                               NaN        New Julie   
3         0932 Gomez Drives\nLeefort, MD 46879-3166              NaN   
4                                               NaN  East Sharonstad   

  address.state             address.street address.zip         dob  \
0            RI  86314 David Pass Apt. 211       83973  1971-06-30   
1           NaN                        NaN         NaN  1965-09-09   
2            UT          6676 Young Square       73125  1993-04-12   
3           NaN                        NaN         NaN  1977-04-14   
4            ME            158 Smith Vista       42483  1970-03-19   

                    email                                id name  \
0       opark@hotmail.com  01d68a4c598a45559c06f4df0b3d82cb  NaN   
1    sperez@armstrong.com  876ff718291d4397bb1e0477ceee6ad9  NaN   
2        uortiz@gmail.com  81753097bf7e4e2085982f422bdb9cda  NaN   
3   palmerdiane@yahoo.com  2c2f7154b80f40ca80d08c5adc54ea45  NaN   
4  nancymaxwell@gmail.com  4f5263f339694d068e17ee7fdbb852b8  NaN   

  name.firstname name.lastname name.middlename                phone  \
0        Cynthia        Dawson          Claire         624-869-4610   
1         Tamara         Myers             NaN       1-594-462-7759   
2          Jamie     Alexander             NaN     472.218.5065x389   
3         Angela        Garcia          Alexis  1-663-109-4460x1080   
4       Jennifer     Rodriguez             NaN         233-423-3823   

           record_date          ssn  
0  2006-07-08T09:02:13  xxx-xx-2412  
1  2009-03-28T20:22:57  xxx-xx-8025  
2  2016-08-30T20:31:39  xxx-xx-0568  
3  2001-02-15T18:50:35  xxx-xx-9825  
4  2014-06-21T14:36:01  xxx-xx-9104

In [304]:
# Last 5 rows
df.tail(5)

address       address.city  \
149995  52805 Higgins Harbor\nJessicamouth, SC 83621-8672                NaN   
149996                                                NaN     Melissachester   
149997      8772 Evans Mill\nThomasborough, AK 26835-9457                NaN   
149998                                                NaN  West Heathermouth   
149999        3178 Mary Plains\nKeithville, VT 34029-5102                NaN   

       address.state       address.street address.zip         dob  \
149995           NaN                  NaN         NaN  1948-08-18   
149996            GA    231 Johnson Point       41818  1988-10-28   
149997           NaN                  NaN         NaN         NaN   
149998            UT  85861 Schmidt Locks       17785  1977-11-10   
149999           NaN                  NaN         NaN  1987-11-17   

                                    email                                id  \
149995                                NaN  91b1d3cbcb794e49a41cfdfe66ff931d   
149996                                NaN  af070fd4476946e6b4755232d8ec8252   
149997       daisy78@johnson-johnson.info  e8cddfe3494e486aae6735e027fd6c8d   
149998           hilledward@gutierrez.net  b47fedaa20bd428fb674aeb8db0ba1cc   
149999  jessicalewis@carroll-martinez.com  6081948439244760865cf6c21ad77248   

                  name name.firstname name.lastname name.middlename  \
149995             NaN        Eduardo         Mejia          Curtis   
149996             NaN          Emily         Solis             NaN   
149997     Anna Walter            NaN           NaN             NaN   
149998  Gary Rodriguez            NaN           NaN             NaN   
149999             NaN           Alex       Johnson             NaN   

                      phone          record_date          ssn  
149995         704-719-8698  2016-09-24T17:00:36  xxx-xx-9311  
149996   (548)476-4285x0779  2014-11-17T01:50:15  xxx-xx-5319  
149997  (926)119-0532x51870  2006-08-16T21:40:19  xxx-xx-0710  
149998         670-853-5186  2012-03-29T18:42:39  xxx-xx-6767  
149999         791-310-0485  2004-05-29T18:44:26  xxx-xx-7983

In [305]:
# Random sample
df.sample(10)

address  address.city  \
29343                40662 Beth Cliff\nNew Lisa, MA 86862           NaN   
14003                                                 NaN  Davischester   
133206  21007 Gilbert Hills Suite 443\nEast Kristina, ...           NaN   
124103  773 Chad Highway Suite 812\nWatkinsfort, AK 68173           NaN   
21868                                                 NaN  South Joshua   
144684                                                NaN   Bryantmouth   
42861   873 Martin Rapid Suite 768\nWest Michaelville,...           NaN   
3903                                                  NaN           NaN   
23631       932 Kimberly Points\nReyesstad, LA 04792-0426           NaN   
77999   24357 Lisa Cove Apt. 728\nNorth Michael, VA 56...           NaN   

       address.state             address.street address.zip         dob  \
29343            NaN                        NaN         NaN  1993-12-27   
14003             VT        881 Jenkins Highway       52857  1976-10-19   
133206           NaN                        NaN         NaN  1982-01-27   
124103           NaN                        NaN         NaN  1962-05-15   
21868             AZ  3286 Atkins Land Apt. 917       59379  1960-06-12   
144684            ME        602 Michael Village  64126-1933  1979-04-05   
42861            NaN                        NaN         NaN  1982-11-29   
3903             NaN                        NaN         NaN  1976-07-17   
23631            NaN                        NaN         NaN  1982-02-09   
77999            NaN                        NaN         NaN  1988-10-08   

                                  email                                id  \
29343        johnsondouglas@hotmail.com  ead8b6edd311494c83769d4d63af559f   
14003             rebecca26@hotmail.com  fc4a1f15bd42443f82ec95354c8cc7f7   
133206     moorerichard@kennedy-fox.biz  253dd94987e7400cbcd61409f900944f   
124103           mirandatara@howell.com  d1e3631e7c5b45408108ef622570c348   
21868   patelmiranda@rodriguez-ryan.com  977889b594f2435c985a26d4c6bf9e7f   
144684             jasonnovak@yahoo.com  fcd209721e3549bba5fb38a424d0288e   
42861    lyonsnicole@woodward-lopez.com  b3ff6d5759c74ffba2a8e1ccd19b24be   
3903                  bryce13@jones.com  ae2e69ab9ece49d789b8243243b525ea   
23631                  tblack@silva.com  fd462a93daf3434b805ea41289847806   
77999                 mason51@gmail.com  9717a9b5c1c84c35a0d2444a51e958ff   

                    name name.firstname name.lastname name.middlename  \
29343                NaN          David      Griffith             NaN   
14003         Anna Silva            NaN           NaN             NaN   
133206    Michael Garcia            NaN           NaN             NaN   
124103    Cynthia Butler            NaN           NaN             NaN   
21868                NaN          Ebony         Parks             NaN   
144684     Steven Miller            NaN           NaN             NaN   
42861                NaN         Sandra       Coleman             NaN   
3903       Natalie Moses            NaN           NaN             NaN   
23631   Rebecca Anderson            NaN           NaN             NaN   
77999       Anthony Ware            NaN           NaN             NaN   

                      phone          record_date          ssn  
29343   (117)902-6025x83700  2014-08-29T23:35:04  xxx-xx-2014  
14003          275-852-3126  2005-08-18T13:42:50  xxx-xx-9188  
133206        (960)598-3259  2006-06-19T03:24:45          NaN  
124103         272-735-7330  2005-10-29T08:49:44  xxx-xx-2969  
21868      +02(4)5928477580  2013-04-12T07:08:34  xxx-xx-3503  
144684                  NaN  2012-05-05T10:14:41  xxx-xx-2887  
42861    508-513-9790x98742  2002-01-05T14:47:41  xxx-xx-4049  
3903            03987357340  2010-05-11T20:06:36  xxx-xx-6378  
23631    405-884-8143x99721  2005-04-22T04:32:46  xxx-xx-5993  
77999                   NaN  2017-01-29T14:28:46  xxx-xx-2601

In [306]:
# displays the total count, unique, top record and the frequency of the top record for all the fields
df.describe()

address address.city  \
count                                              75193        61233   
unique                                             75193        28572   
top     7343 Villegas Knolls\nEast Christopher, CO 80727  New Michael   
freq                                                   1           61   

       address.state               address.street address.zip         dob  \
count          61233                        61233       61233      143874   
unique            59                        61233       54528       17164   
top               NC  42418 Austin View Suite 990       99528  1950-11-13   
freq            1109                            1           5          24   

                   email                                id         name  \
count             130880                            150000        43058   
unique            124093                            150000        35305   
top     qsmith@gmail.com  11fc0c232e374ec5877bb9c959296a44  David Smith   
freq                   9                                 1           20   

       name.firstname name.lastname name.middlename           phone  \
count          105004        105004           43669          140261   
unique            690          1000             689          140260   
top           Michael         Smith         Michael  1-298-341-1025   
freq             2356          2373             977               2   

                record_date          ssn  
count                150000       142444  
unique               149977         9999  
top     2005-12-25T00:33:49  xxx-xx-6568  
freq                      2           32

## Missing Data

In row 1667, we see that there are NaN for fields [(address.city, address.state, address.street, address.zip), (name), (name.middlename) and (ssn)].

There are two types of missing data,
1. Ones that are missing during the collection phase
        In the above example it's the field 'ssn'
        
        
2. These are the missing values that is returned because of json_normalize 
        Here in row 1,'address' is present and (address.city, address.state, address.street, address.zip) is Nan. This can be seen in fields like name and address because of the nested field of json. 'name.firstname' and 'name.lastname' are present but 'name' is Nan. (These NaN values can be ignored)

In [307]:
df[1667:1668]

address address.city  \
1667  346 Hanson Islands Suite 799\nSouth Kristin, W...          NaN   

     address.state address.street address.zip         dob  \
1667           NaN            NaN         NaN  1987-04-17   

                            email                                id name  \
1667  kevin41@larson-watkins.info  3989ac5ada25453099e5c933f56bcc8b  NaN   

     name.firstname name.lastname name.middlename          phone  \
1667          Brian       Hampton             NaN  (931)051-2250   

              record_date  ssn  
1667  2004-06-07T11:40:40  NaN

## --------------------------------------------------------------------------------------------------------------------------------
## Ordered list of columns

In [308]:
print(df.columns.tolist())

['address', 'address.city', 'address.state', 'address.street', 'address.zip', 'dob', 'email', 'id', 'name', 'name.firstname', 'name.lastname', 'name.middlename', 'phone', 'record_date', 'ssn']


## --------------------------------------------------------------------------------------------------------------------------------
## Percentage of records that each field contains

In [309]:
# df.count() returns the count of all the fields and perRec is the percentage records 
freqCount = df.count()
perRec = 100 * (freqCount/rows)
perRec = perRec.map('{:,.1f}%'.format)
print(perRec)

address             50.1%
address.city        40.8%
address.state       40.8%
address.street      40.8%
address.zip         40.8%
dob                 95.9%
email               87.3%
id                 100.0%
name                28.7%
name.firstname      70.0%
name.lastname       70.0%
name.middlename     29.1%
phone               93.5%
record_date        100.0%
ssn                 95.0%
dtype: object


## --------------------------------------------------------------------------------------------------------------------------------
## 5 most common values 

In [310]:
# df[i].value_counts() returns the freuqency of distinct values in the field i
for i in df.columns:
    print(df[i].value_counts()[:5])
    print()

7343 Villegas Knolls\nEast Christopher, CO 80727             1
Unit 0496 Box 2344\nDPO AP 23608-8659                        1
5234 Lopez Motorway\nWest Christinemouth, MH 51591-1867      1
PSC 6495, Box 1176\nAPO AE 13397-7432                        1
2745 Swanson Rapids Apt. 768\nNew Markview, IN 21701-3708    1
Name: address, dtype: int64

New Michael      61
Lake Michael     58
East Michael     57
North Michael    56
Port Michael     56
Name: address.city, dtype: int64

NC    1109
DC    1106
MI    1097
MD    1095
OH    1085
Name: address.state, dtype: int64

42418 Austin View Suite 990        1
3049 Abigail Extension Apt. 243    1
848 Paul Ramp Suite 873            1
658 Garcia Plaza Suite 559         1
73866 Pamela Lake                  1
Name: address.street, dtype: int64

99528    5
50032    5
98018    5
53097    5
17086    5
Name: address.zip, dtype: int64

1950-11-13    24
1963-01-25    21
1963-11-30    20
1964-06-11    20
1989-08-30    20
Name: dob, dtype: int64

qsmith@gmail.

## --------------------------------------------------------------------------------------------------------------------------------
## Distinct first names in the data set

Two dataframes: namesDF <- name field from df, firstName <- name.firstname field from df
    
Steps:
1. Create the above mentioned dataframes and rename the column name to name.firstname in namesDF
2. Use namesDF.update(firstName) to modifiy the namesDF in place using non-NA values from firstName
3. Drop any nan values from namesDF (These are the original missing names)
4. Use re.match() to match the first part of the name field and remove the name prefixes like Mr., Mrs. or Dr.
       For example: Dr. Morgan Morris is returned as Morgan
5. Use fName['name.firstname'].nunique() to retrieve the number of unique values from fName

In [311]:
namesDF = df['name'].to_frame()
namesDF.rename(columns={'name': 'name.firstname'}, inplace=True)
firstName = df['name.firstname'].to_frame()

namesDF.update(firstName)
namesDF = namesDF.dropna()
namesDF.head(15)

name.firstname
0             Cynthia
1              Tamara
2               Jamie
3              Angela
4            Jennifer
5               Keith
6               Shane
7             Charles
8       Morgan Morris
9   Bernard Wilkerson
10             Amanda
11             Adrian
12         Casey Haas
13            Tiffany
14     Barbara Harris

In [312]:
# applymap(func) where func matches only the first part of the name field and removes the Prefix of the names
fName = namesDF.applymap(lambda name: re.match(r'^(?:(?:Mr|Mrs|Dr)\.\s)?([\S]+)\s?', str(name)).group(1))
fName.head(15)

name.firstname
0         Cynthia
1          Tamara
2           Jamie
3          Angela
4        Jennifer
5           Keith
6           Shane
7         Charles
8          Morgan
9         Bernard
10         Amanda
11         Adrian
12          Casey
13        Tiffany
14        Barbara

In [313]:
distNames = fName['name.firstname'].nunique()
print("Number of distinct Names: ", distNames)

Number of distinct Names:  692


## --------------------------------------------------------------------------------------------------------------------------------
## Distinct street names in the data set

Two dataframes: addr <- address field from df, addSt <- address.street field from df
    
Steps:
1. Create the above mentioned dataframes 
2. Create addrDF by applying applymap(func) where func strips the first line of the value in the address field and    rename the column name to address.street
2. Use addrDF.update(addSt) to modifiy the addrDF in place using non-NA values from addSt
3. Call regexMatch using applymap() on values of addrDF.
4. regexMatch function
    returns None for those values which starts with PSC|Unit|USNS|USNV|USS|USCGC as these are Postal Service Center     units and others are ship prefixes
    returns the street address for the rest by stripping off the numbers and Apt/Suite number from the values
        For example: 86314 David Pass Apt. 211 is returned as David Pass
5. stName is converted to a set so that the values are distinct and nan and None values are removed from the set.
6. len(stName) is returned which is the length of the distinct Street names in the dataset

In [314]:
addr = df['address'].to_frame()
addSt = df['address.street'].to_frame()

# Split the first line of the address field values
addrDF = addr.applymap(lambda ad: str(ad).split("\n")[0])
addrDF.rename(columns={'address': 'address.street'}, inplace=True)

addrDF.update(addSt)
addrDF.head(15)

address.street
0        86314 David Pass Apt. 211
1           20722 Coleman Villages
2                6676 Young Square
3                0932 Gomez Drives
4                  158 Smith Vista
5                 461 Knapp Unions
6      9524 Danielle Burg Apt. 849
7                407 Patrick Shoal
8      3697 Mills Estates Apt. 499
9                              nan
10   01737 Hailey Drives Suite 056
11         395 Anna Roads Apt. 613
12              56833 Baker Branch
13  045 Bennett Motorway Suite 404
14     908 Matthew Ports Suite 313

In [315]:
def regexStMatch(street):
    # returns None when the address starts with PSC|Unit|USNS|USNV|USS|USCGC
    if re.match(r'^(?:PSC|Unit|USNS|USNV|USS|USCGC)', str(street)):
        return
    # returns the streetName from the address field by stripping numbers and Apt/Suite
    return re.match(r'^(?:\d+\s+)?(.*?)(?:\s+(?:Apt\.|Suite)\s+\d+)?$', str(street)).group(1)

stName = addrDF.applymap(regexStMatch)
stName.head(15)

address.street
0         David Pass
1   Coleman Villages
2       Young Square
3       Gomez Drives
4        Smith Vista
5       Knapp Unions
6      Danielle Burg
7      Patrick Shoal
8      Mills Estates
9                nan
10     Hailey Drives
11        Anna Roads
12      Baker Branch
13  Bennett Motorway
14     Matthew Ports

In [316]:
# stName is a set so that only distinct values are present
stName = set(stName.to_dict()['address.street'].values())

# Drop those rows with 'nan' and None type
stName.remove('nan')
stName.remove(None)

print("Distinct Street Names: ", len(stName))

Distinct Street Names:  79999


## --------------------------------------------------------------------------------------------------------------------------------
## 5 most common US area codes in phone number field

US area codes(first 3 digits after the country code)

Treated as US phone numbers if it begins with:
    1. (Country code) 1- or 1 or 1. followed by the regex of 2
        Ex: 1-(420)5089422
            1-420-508-9422
            1(420)5089422
            14205089422
            1.(420)5089422
            1.4205089422\
          
    2. \([1-9]\d\d\)
        Ex: (420)5089422
        
    3. [1-9]\d\d 
        Ex:  4205089422
        
       (2 and 3 kind of values could belong to any Country but for this analysis, its assumed to be US numbers) 

Steps:
1. phoneDF is created from df['phone']
2. Call getAreacode using apply() on values of phoneDF.
4. getAreacode function
    returns the stripped numbers (3 digits which are area codes) by matching the regular expression.
        For ex: 1(420)5089422 returned as 420
                +39(2)5142930962 is passed
                4205089422 returned as 420
5. NaN values are dropped
6. Using value_counts()[:5] on areaCodes we get the 5 most common US area codes

In [317]:
phoneDF = df['phone']

def getAreacode(number):
    try:
        # returns the first 3 digit of a phone number that either starts with 1/1-/(/one of the [1-9]
        return re.match(r'(?:1-|1|1\.)?\(?([1-9]\d\d)\)?', str(number)).group(1)
    except AttributeError:
        pass
# Non-null phone number records

areaCodes = phoneDF.apply(getAreacode)
areaCodes.dropna(inplace=True)
print("5 most common US area codes: ")
areaCodes.value_counts()[:5]

5 most common US area codes: 


947    154
488    148
913    148
740    146
111    145
Name: phone, dtype: int64

## --------------------------------------------------------------------------------------------------------------------------------
## Future Work

1. Most common domain used in email address ('gmail', 'yahoo', 'hotmail' etc)
2. How many people were born in a selected year?
3. What is the most popular name of the year?
4. Find the distinct lastName and n number of most common lastNames.
5. Populous state, street and the most common name in that state, street.
6. Most Common and Least common names, state, city etc
7. Number of records recorded over the years.
8. Number of missing SSN and their age
9. What is the age of the population and in which age group do they fall? (min, max and in ranges)